vault/ssorolecredentialsprovider.go

```diff
@@ -0,0 +1,272 @@
+
+	return creds, nil
+}
+
+type SSOClientCredentials struct {
+	ID         string
+	Secret     string
+	Expiration time.Time
+}
+
+type SSOAccessToken struct {
+	Token      string
+	Expiration time.Time
+}
+
+type SSOOIDCProvider struct {
+	OIDCClient *ssooidc.SSOOIDC
+	Keyring    *CredentialKeyring
+	StartURL   string
+}
+
+func (p *SSOOIDCProvider) GetAccessToken() (*SSOAccessToken, error) {
+	var (
+		creds = &struct {
+			Token  *SSOAccessToken
+			Client *SSOClientCredentials
+		}{
+			Client: &SSOClientCredentials{},
+			Token:  &SSOAccessToken{},
+		}
+		credsUpdated bool
+	)
+
+	item, err := p.Keyring.Keyring.Get(p.StartURL)
+	if err != nil && err != keyring.ErrKeyNotFound {
+		return nil, err
+	}
+
+	if item.Data != nil {
+		if err = json.Unmarshal(item.Data, &creds); err != nil {
+			return nil, fmt.Errorf("Invalid data in keyring: %v", err)
+		}
+	}
+
+	if creds.Client.Expiration.Before(time.Now()) {
+		creds.Client, err = p.registerNewClient()
+		if err != nil {
+			return nil, err
+		}
+		log.Printf("Created new SSO client for %s (expires at: %s)", p.StartURL, creds.Client.Expiration.String())
+		credsUpdated = true
+	}
+
+	if creds.Token.Expiration.Before(time.Now()) {
+		creds.Token, err = p.createClientToken(creds.Client)
+		if err != nil {
+			return nil, err
+		}
+		log.Printf("Created new SSO access token for %s (expires at: %s)", p.StartURL, creds.Token.Expiration.String())
+		credsUpdated = true
+	}
+
+	if credsUpdated {
+		bytes, err := json.Marshal(creds)
+		if err != nil {
+			return nil, err
+		}
+		err = p.Keyring.Keyring.Set(keyring.Item{
+			Key:                         p.StartURL,
+			Label:                       fmt.Sprintf("aws-vault (%s)", p.StartURL),
+			Data:                        bytes,
+			KeychainNotTrustApplication: true,
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return creds.Token, nil
+}
+
+func (p *SSOOIDCProvider) registerNewClient() (*SSOClientCredentials, error) {
+	c, err := p.OIDCClient.RegisterClient(&ssooidc.RegisterClientInput{
+		ClientName: aws.String(ssoClientName),
+		ClientType: aws.String(ssoClientType),
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &SSOClientCredentials{
+		ID:         aws.StringValue(c.ClientId),
+		Secret:     aws.StringValue(c.ClientSecret),
+		Expiration: time.Unix(aws.Int64Value(c.ClientSecretExpiresAt), 0),
+	}, nil
+}
+
+func (p *SSOOIDCProvider) createClientToken(creds *SSOClientCredentials) (*SSOAccessToken, error) {
+	auth, err := p.OIDCClient.StartDeviceAuthorization(&ssooidc.StartDeviceAuthorizationInput{
+		ClientId:     aws.String(creds.ID),
+		ClientSecret: aws.String(creds.Secret),
+		StartUrl:     aws.String(p.StartURL),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	fmt.Printf(authorizationTemplate, aws.StringValue(auth.VerificationUriComplete))
```

Comment:
If we dump this to stdout with `fmt.Printf` it may interfere with the output of whatever command you're executing... can we get this output to stderr, preferably through kingpin somehow? Alternatively should we be using the `prompt` driver to achieve this?

Reply:
Great catch! Since `prompt` is only used for requesting user input (which we are not doing here), and `kingpin` is not currently used anywhere outside the `cli/` page right now, I think it makes most sense to use `Fprintf(os.Stderr, ...)` here.